File: NumericalIntegral/comsimp.jl
"""
    comsimp(func::Function, a::Real, b::Real, n::Integer)

Numerically approximate the integral of func from `a` to `b` using Simpson's composite rule.

See also [`simpson`], [`adapsimp`].

# Arguments

* `func::Function`: `func` is the integrand.
* `a::Real`: `a` is the lower limit of the integration by default.
* `b::Real`: `b` is the upper limit of the integration by default. 

If `a = b`, throw a Warning and return 0; if `a > b`, swap `a` and `b` to ensure `h = (b - a)/n > 0`.

* `n::Integer`: `n` is the number of intervals in equal parts and `n` is greater than or equal to 2.

# Outputs

* `s::Float64`: `s` is the approximation of integral.

# Examples

```julia-repl
julia> comsimp(log, 1, 2, 1)
Error! The Simpson's composite rule requires the interval to be divided into at
least two parts.

julia> comsimp(cos, 0, 0, 4)
Warning! The limits of integration are equal!
0

julia> comsimp(sin, 0, pi/2, 4)
1.0001345849741938

julia> comsimp(x->exp(-x^2), 1, 0, 4)
0.7468553797909873
```
"""
function comsimp(func::Function, a::Real, b::Real, n::Integer)
    if n < 2
        @error "n is greater than or equal to 2!"
        return
    end
    if a == b
        @warn "The limits of integration are equal!"
        return 0
    elseif a > b
        a, b = b, a
    end

    h = (b - a) / n
    s0 = func(a) + func(b)
    s1 = 0      # summation of f(x_{2k-1})
    s2 = 0      # summation of f(x_{2k})
    for k = 1:n-1
        x = a + k * h
        if rem(k, 2) == 0
            s2 = s2 + func(x)
        else
            s1 = s1 + func(x)
        end
    end
    s = h * (s0 + 4 * s1 + 2 * s2) / 3
    return s
end

File: NumericalIntegral/comsimp.m
function s = comsimp(func, a, b, n)
%COMSIMP - 复合 Simpson 公式数值求解函数 func 在区间 [a, b] 上的定积分.
%
% 语法: s = comsimp(func, a, b, n)
%
% 输入:
%   func - 被积函数. func 是一个函数句柄或者匿名函数.
%   a - double. a 默认为积分下限.
%   b - double. b 默认为积分上限.
%       如果 a = b, 抛出积分上下限相等的警告并赋 s = 0.
%       如果 a > b, 交换 a 和 b 的值保证 h = (b - a) / n > 0.
%   n - integer. n 是积分区间等分数. n 不小于 2.
%
% 输出:
%   s - double. s 是积分的近似值.
%
% 示例:
% > comsimp(@log, 1, 2, 1)
% 错误! n 不小于 2!
% > comsimp(@cos, 0, 0, 4)
% 警告! 积分上下限相等!
% ans =
%   0
% > comsimp(@sin, 0, pi/2, 4)
% ans =
%   1.000134584974194
% > comsimp(@(x) exp(-x^2), 1, 0, 4)
% ans =
%   0.746855379790987
%
% 另见: simpson, adapsimp
%
%
    if n < 2
        error("错误! n 不小于 2!")
    end
    if a == b
        disp("警告! 积分上下限相等!")
        s = 0;
        return;
    elseif a > b
        c = a;
        a = b;
        b = c;
    end

    h = (b - a) / n;
    s0 = func(a) + func(b);
    s1 = 0;     % summation of f(x_{2k-1})
    s2 = 0;     % summation of f(x_{2k})
    for k = 1:n-1
        x = a + k * h;
        if rem(k , 2) == 0
            s2 = s2 + func(x);
        else
            s1 = s1 + func(x);
        end
    end
    s = h * (s0 + 4 * s1 + 2 * s2) / 3;
end
